filestore/filepath.go

```diff
@@ -0,0 +1,399 @@
+	err := fp.parsePath(fullPath)
+	if err != nil {
+		return fmt.Errorf("dir: %v", err)
+	}
+	// TODO: consider removing the final piece of the path if it's a file
+	fp.isDir = true
+	return nil
+}
+
+func (fp *FilePath) checkSchemes(scheme string) error {
+	for _, s := range ValidSchemes {
+		if s == scheme {
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid scheme '%s', must be one of %v", scheme, ValidSchemes)
+}
+
+func (fp *FilePath) parsePath(fullPath string) error {
+	// Parse the URI into a url.URL object.
+	u, err := url.Parse(fullPath)
+	if err != nil {
+		return fmt.Errorf("could not parse fullpath '%s': %v", fullPath, err)
+	}
+
+	// Extract the bucket and path components from the URI.
+	bucket := u.Host
+	path := strings.TrimPrefix(u.Path, "/")
+
+	err = fp.checkSchemes(u.Scheme)
+	if err != nil {
+		return err
+	} else {
+		fp.scheme = u.Scheme
+	}
+
+	fp.bucket = bucket
+	fp.key = path
+	return nil
+}
+
+func (fp *FilePath) IsValid() bool {
+	return fp.isValid
+}
+
+func (fp *FilePath) Validate() error {
+	return fmt.Errorf("not implemented")
+}
+
+type S3Filepath struct {
+	FilePath
+}
+
+func (s3 *S3Filepath) Validate() error {
+	if s3.scheme != "s3://" && s3.scheme != "s3a://" {
+		return fmt.Errorf("invalid scheme '%s', must be 's3:// or 's3a://'", s3.scheme)
+	}
+	if s3.bucket == "" {
+		return fmt.Errorf("bucket cannot be empty")
+	} else {
+		s3.bucket = strings.Trim(s3.bucket, "/")
+	}
+	if s3.key == "" {
+		return fmt.Errorf("key cannot be empty")
+	} else {
+		s3.key = strings.Trim(s3.key, "/")
+	}
+
+	s3.isValid = true
+	return nil
+}
+
+func (s3 *S3Filepath) PathWithBucket() string {
+	return fmt.Sprintf("%s%s/%s", s3.scheme, s3.bucket, s3.key)
+}
+
+type AzureFilepath struct {
+	StorageAccount string
+	FilePath
+}
+
+func (azure *AzureFilepath) PathWithBucket() string {
+	return fmt.Sprintf("abfss://%s@%s.dfs.core.windows.net/%s", azure.bucket, azure.StorageAccount, azure.key)
+}
+
+// **NOTE**: Due to Azure Blob Storage's unique URI format, we need to re-implement this method
+// on the derived type to ensure we can properly handle the `bucket` field.
+func (azure *AzureFilepath) ParseFilePath(fullPath string) error {
+	u, err := url.Parse(fullPath)
+	if err != nil {
+		return fmt.Errorf("could not parse full path '%s': %v", fullPath, err)
+	}
+	// Our scheme is the protocol + "://", so we need to suffix the scheme with "://"
+	// to ensure the comparison works.
+	err = azure.FilePath.checkSchemes(fmt.Sprintf("%s://", u.Scheme))
+	if err != nil {
+		return err
+	} else {
+		azure.FilePath.scheme = u.Scheme
+	}
+	azure.FilePath.scheme = u.Scheme
+	azure.FilePath.bucket = u.User.String()              // The container will be in the User field due to the format <scheme>://<container>@<storage_account>
+	azure.StorageAccount = strings.Split(u.Host, ".")[0] // The host will be in the format <storage_account>.dfs.core.windows.net
+	azure.FilePath.key = strings.TrimPrefix(u.Path, "/")
+	azure.FilePath.isDir = false
```

Comment:
Is there a reason why this is false? can you pass a folder as key?

Reply:
The idea here is that you're creating a `FilePath` instance that points to a file; otherwise, you'd use `CreateDirPath`.